configuration/outbound.md

```diff
@@ -0,0 +1,201 @@
+  "id": "abc-1234",
+  "type": "data_record",
+  "form": "referral",
+  "reported_date": 1555069530966,
+  "fields": {
+    "patient_id": "12345",
+    "patient_temp_deg_F": 100,
+    "danger_signs": ["V", "BL", "D", "IG"]
+  },
+  "contact": {
+    "_id": "def-5678",
+    "name": "Jane CHW",
+    "parent": {
+      "_id": "hij-9012",
+      "name": "..."
+    }
+  }
+}
+```
+
+The following mapping configuration:
+```json
+{
+  "mapping": {
+    "patient_id": "doc.fields.patient_id",
+    "chw_id": "doc.contact._id",
+    "event.temp_c": {
+      "expr": "((doc.fields.patient_temp_deg_F - 32) * (5.0/9.0)).toFixed(2)"
+    },
+    "event.in_danger": {
+      "expr": "doc.fields.danger_signs.length >= 3 ? true : undefined",
+      "optional": true
+    },
+    "event.additional_notes": {
+      "path": "doc.fields.notes",
+      "optional": true
+    }
+  }
+}
+```
+
+Would create the following JSON payload to send:
+```json
+{
+  "patient_id": "12345",
+  "chw_id": "def-5678",
+  "event": {
+    "temp_c": 37.78,
+    "in_danger": true
+  }
+}
+```
+
+This example makes a few points:
+ - The report that is being used to generate the outbound push is referenced as `doc` in both `path` and `expr` properties
+ - To define a property that is itself an object, you can make the mapping key a JSON path.
+ - If you define a property as `optional`, it won't exist at all in the payload if the resulting value is `undefined`, either because that is result of executing the `expr`, or the `path` doesn't exist. Note that if the `event.in_danger` expression was instead `doc.fields.danger_signs.length >= 3` the property `in_danger` would always exist and would either be `true` or `false`
+
+Other notes:
+ - Your report will be hydrated before being passed to the mapper. This gives you access to the contact and its parents
+ - JSON paths that may have undefined properties need to be dealt with differently depending on if you are using a `path` or an `expr`. Given `doc.foo.bar.smang` as a path where any of those properties may not exist in the doc:
+  - If you're using `path` just use the path as is, if any part of the path is `undefined` the resulting value will safely be `undefined`
+  - However, in `expr` you **do** need to handle this situation: `doc.foo && doc.foo.bar && doc.foo.bar.smang`
+- If any of your `expr` expressions throw an exception (for example because you didn't handle potentially `undefined` properties as noted above) your push will fail
+- If any of your `path` declarations result in an `undefined` value and you have not also declared that property optional your push will fail
+
+## Credentials
+
+To securely store credentials, we'll be using CouchDB's config storage, as this is a convenient place that only CouchDB admins can access.
+
+Passwords are stored under the `medic-credentials` section, under the key declared in config.
+
+In the following example:
+
+```json
+{
+  "destination": {
+    "base_url": "https://example.com",
+    "auth": {
```

Comment:
Is it useful to abbreviate these interfaces? Like "auth" instead of "authentication" and "expr" instead of "expression"?

Reply:
In the case of `expr` I picked it specifically because it already exists throughout config to mean the same thing with the same constraints (a JS expression that has the doc in context).

I can change auth to authentication if you want though.